File: web/src/sim/buildings/BuildingType.ts
import { divF32, f32, mulF32 } from '~/sim/determinism/float32'

export enum BuildingType {
  ArmyBase = 0,
  ArmsIndustry = 1,
  AirBase = 2,
  NavalBase = 3,
  RecruitingOffice = 4,
  MilitaryHospital = 5,
  UndergroundBunkers = 6,
  SecretWeaponsLab = 7,
}

export const BUILDING_COUNT = 8
export const MAX_BUILDING_LEVEL = 5

/* Urutan tetap menggantikan Enum.GetValues: mengiterasi kunci objek enum akan
   ikut memuat pemetaan balik namanya dan urutannya tidak dijamin. */
export const ALL_BUILDING_TYPES: readonly BuildingType[] = [
  BuildingType.ArmyBase,
  BuildingType.ArmsIndustry,
  BuildingType.AirBase,
  BuildingType.NavalBase,
  BuildingType.RecruitingOffice,
  BuildingType.MilitaryHospital,
  BuildingType.UndergroundBunkers,
  BuildingType.SecretWeaponsLab,
]

const NAMES: readonly string[] = [
  'Army Base',
  'Arms Industry',
  'Air Base',
  'Naval Base',
  'Recruiting Office',
  'Military Hospital',
  'Underground Bunkers',
  'Secret Weapons Lab',
]

/* Tekanan slot itulah yang membuat kota jadi sekumpulan keputusan alih-alih
   daftar centang: delapan bangunan ada, satu kota memuat paling banyak tujuh,
   sehingga spesialisasi dipaksa, bukan dianjurkan. */
export function slotsFor(population: number): number {
  const slots = 2 + Math.trunc(divF32(population, 2))
  return Math.min(Math.max(slots, 2), 7)
}

export function nameOf(type: BuildingType): string {
  return NAMES[type] ?? 'Secret Weapons Lab'
}

export function requiresCoast(type: BuildingType): boolean {
  return type === BuildingType.NavalBase
}

/* Bonus keluaran dari Arms Industry, mengikuti kemajuan Conflict of Nations
   yang dicatat riset ekonomi. */
export function productionBonus(type: BuildingType, level: number): number {
  return type === BuildingType.ArmsIndustry ? mulF32(level, f32(0.1)) : 0
}

export function manpowerBonus(type: BuildingType, level: number): number {
  return type === BuildingType.RecruitingOffice ? mulF32(level, f32(0.05)) : 0
}

export function moraleBonus(type: BuildingType, level: number): number {
  if (type !== BuildingType.UndergroundBunkers) return 0

  switch (level) {
    case 1:
      return f32(0.05)
    case 2:
      return f32(0.1)
    case 3:
      return f32(0.2)
    case 4:
      return f32(0.35)
    default:
      return f32(0.5)
  }
}

File: sim/NationRise.Core/Buildings/BuildingType.cs
namespace NationRise.Core.Buildings;

public enum BuildingType : byte
{
    ArmyBase = 0,
    ArmsIndustry = 1,
    AirBase = 2,
    NavalBase = 3,
    RecruitingOffice = 4,
    MilitaryHospital = 5,
    UndergroundBunkers = 6,
    SecretWeaponsLab = 7,
}

public static class BuildingInfo
{
    public const int Count = 8;
    public const int MaxLevel = 5;

    /* Slot pressure is what makes a city a set of decisions rather than a
       checklist: nine buildings exist, a city fits at most seven, so
       specialising is forced rather than encouraged. */
    public static int SlotsFor(float population) =>
        Math.Clamp(2 + (int)(population / 2f), 2, 7);

    public static string NameOf(BuildingType type) => type switch
    {
        BuildingType.ArmyBase => "Army Base",
        BuildingType.ArmsIndustry => "Arms Industry",
        BuildingType.AirBase => "Air Base",
        BuildingType.NavalBase => "Naval Base",
        BuildingType.RecruitingOffice => "Recruiting Office",
        BuildingType.MilitaryHospital => "Military Hospital",
        BuildingType.UndergroundBunkers => "Underground Bunkers",
        _ => "Secret Weapons Lab",
    };

    public static bool RequiresCoast(BuildingType type) => type == BuildingType.NavalBase;

    /* Output bonus from Arms Industry, matching the Conflict of Nations
       progression the economy research recorded. */
    public static float ProductionBonus(BuildingType type, int level) =>
        type == BuildingType.ArmsIndustry ? level * 0.10f : 0f;

    public static float ManpowerBonus(BuildingType type, int level) =>
        type == BuildingType.RecruitingOffice ? level * 0.05f : 0f;

    public static float MoraleBonus(BuildingType type, int level) => type switch
    {
        BuildingType.UndergroundBunkers => level switch
        {
            1 => 0.05f, 2 => 0.10f, 3 => 0.20f, 4 => 0.35f, _ => 0.50f,
        },
        _ => 0f,
    };
}
